sqlmesh/integrations/github/cicd/controller.py

```diff
@@ -0,0 +1,717 @@
+
+    from sqlmesh.core.config import Config
+    from sqlmesh.core.snapshot import Snapshot
+
+logger = logging.getLogger(__name__)
+
+
+class PullRequestInfo(PydanticModel):
+    """Contains information related to a pull request that can be used to construct other objects/URLs"""
+
+    owner: str
+    repo: str
+    pr_number: int
+
+    @property
+    def full_repo_path(self) -> str:
+        return "/".join([self.owner, self.repo])
+
+    @classmethod
+    def create_from_pull_request_url(cls, pull_request_url: str) -> PullRequestInfo:
+        _, _, _, _, owner, repo, _, pr_number = pull_request_url.split("/")
+        return cls(
+            owner=owner,
+            repo=repo,
+            pr_number=pr_number,
+        )
+
+
+class GithubCheckStatus(str, Enum):
+    QUEUED = "queued"
+    IN_PROGRESS = "in_progress"
+    COMPLETED = "completed"
+
+    @property
+    def is_queued(self) -> bool:
+        return self == GithubCheckStatus.QUEUED
+
+    @property
+    def is_in_progress(self) -> bool:
+        return self == GithubCheckStatus.IN_PROGRESS
+
+    @property
+    def is_completed(self) -> bool:
+        return self == GithubCheckStatus.COMPLETED
+
+
+class GithubCheckConclusion(str, Enum):
+    SUCCESS = "success"
+    FAILURE = "failure"
+    NEUTRAL = "neutral"
+    CANCELLED = "cancelled"
+    TIMED_OUT = "timed_out"
+    ACTION_REQUIRED = "action_required"
+    SKIPPED = "skipped"
+
+    @property
+    def is_success(self) -> bool:
+        return self == GithubCheckConclusion.SUCCESS
+
+    @property
+    def is_failure(self) -> bool:
+        return self == GithubCheckConclusion.FAILURE
+
+    @property
+    def is_neutral(self) -> bool:
+        return self == GithubCheckConclusion.NEUTRAL
+
+    @property
+    def is_cancelled(self) -> bool:
+        return self == GithubCheckConclusion.CANCELLED
+
+    @property
+    def is_timed_out(self) -> bool:
+        return self == GithubCheckConclusion.TIMED_OUT
+
+    @property
+    def is_action_required(self) -> bool:
+        return self == GithubCheckConclusion.ACTION_REQUIRED
+
+    @property
+    def is_skipped(self) -> bool:
+        return self == GithubCheckConclusion.SKIPPED
+
+
+class AffectedEnvironmentModel(PydanticModel):
+    model_name: str
+    view_name: str
+    intervals: Intervals
+    change_category: SnapshotChangeCategory
+    interval_unit: t.Optional[IntervalUnit]
+
+    @classmethod
+    def from_snapshot(cls, snapshot: Snapshot) -> AffectedEnvironmentModel:
+        assert snapshot.change_category
+        return cls(
+            model_name=snapshot.model.name,
+            view_name=snapshot.model.view_name,
+            intervals=snapshot.dev_intervals
+            if snapshot.change_category.is_forward_only
+            else snapshot.intervals,
+            change_category=snapshot.change_category,
+        )
+
+    @property
+    def formatted_loaded_intervals(self) -> str:
+        merged_inclusive_intervals = [
```

Comment:
this looks copied from plan/definition.py, can we reuse MissingInterval here?

Reply:
See https://github.com/TobikoData/sqlmesh/pull/810/commits/59326ebd90a572466b473f7708ba22ec8c050469 and https://github.com/TobikoData/sqlmesh/pull/810/commits/cf90a31bb9ffc46b7727de79221661702c5df8af

This does change the normal formatting to use ` - ` between start and end but I think this looks better anyways.